File: mats-websockets/client/javascript/test/unit.js
// Register as an UMD module - source: https://github.com/umdjs/umd/blob/master/templates/commonjsStrict.js
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define(['chai', 'sinon', 'ws', 'mats'], factory);
    } else if (typeof exports === 'object' && typeof exports.nodeName !== 'string') {
        // CommonJS
        const chai = require('chai');
        const sinon = require('sinon');
        const ws = require('ws');
        const mats = require('../lib/MatsSocket');

        factory(chai, sinon, ws, mats);
    } else {
        // Browser globals
        factory(chai, sinon, WebSocket, mats);
    }
}(typeof self !== 'undefined' ? self : this, function (chai, sinon, ws, mats) {
    const MatsSocket = mats.MatsSocket;

    describe('MatsSocket unit-tests', function () {
        describe('constructor', function () {
            it('Should fail on no arg invocation', function () {
                chai.assert.throws(() => new MatsSocket());
            });

            it('Should fail if appVersion and urls are missing', function () {
                chai.assert.throws(() => new MatsSocket("Test"));
            });

            it('Should fail if urls are missing', function () {
                chai.assert.throws(() => new MatsSocket("Test", "1.0"));
            });
            it('Should not invoke the WebSocket constructor', function () {
                const callback = sinon.spy(ws);
                new MatsSocket("Test", "1.0", ["ws://localhost:8080/"]);
                chai.assert(!callback.called);
                sinon.reset();
            });
        });

        describe('authorization', function () {
            // :: Make mock WebSocket
            const webSocket = {};
            // Make send function:
            webSocket.send = function (payload) {
                JSON.parse(payload).forEach(({t, cmseq}, idx) => {
                    if (t === 'HELLO') {
                        setTimeout(() => {
                            webSocket.onmessage({data: JSON.stringify([{t: "WELCOME"}])});
                        }, idx);
                    }
                    if (cmseq !== undefined) {
                        setTimeout(() => {
                            webSocket.onmessage({data: JSON.stringify([{t: "RECEIVED", cmseq: cmseq, st: "ACK"}])});
                        }, idx);
                    }
                });
            };
            webSocket.close = function() {};

            // Make 'ononpen' property, which is set twice by MatsSocket: Once when it waits for it to open, and when this happens, it is "unset" (set to undefined)
            Object.defineProperty(webSocket, "onopen", {
                set(callback) {
                    // When callback is set, immediately invoke it on next tick (fast opening times on this mock WebSockets..!)
                    if (callback !== undefined) {
                        setTimeout(() => callback({}), 0);
                    }
                }
            });

            it('Should invoke authorization callback before making calls', async function () {
                const matsSocket = new MatsSocket("Test", "1.0", ["ws://localhost:8080/"], () => webSocket);

                let authCallbackCalled = false;

                matsSocket.setAuthorizationExpiredCallback(function (event) {
                    authCallbackCalled = true;
                    matsSocket.setCurrentAuthorization("Test", Date.now() + 20000, 0);
                });
                await matsSocket.send("Test.authCallback", "SEND_" + matsSocket.id(6), {});

                chai.assert(authCallbackCalled);
            });

            it('Should not invoke authorization callback if authorization present', async function () {
                const matsSocket = new MatsSocket("Test", "1.0", ["ws://localhost:8080/"], () => webSocket);

                let authCallbackCalled = false;
                matsSocket.setCurrentAuthorization("Test", Date.now() + 20000, 0);
                matsSocket.setAuthorizationExpiredCallback(function (event) {
                    authCallbackCalled = true;
                });
                await matsSocket.send("Test.authCallback", "SEND_" + matsSocket.id(6), {});

                chai.assert(!authCallbackCalled);
            });

            it('Should invoke authorization callback when expired', async function () {
                const matsSocket = new MatsSocket("Test", "1.0", ["ws://localhost:8080/"], () => webSocket);

                let authCallbackCalled = false;

                matsSocket.setCurrentAuthorization("Test", Date.now() - 20000, 0);
                matsSocket.setAuthorizationExpiredCallback(function (event) {
                    authCallbackCalled = true;
                    matsSocket.setCurrentAuthorization("Test", Date.now() + 20000, 0);
                });
                await matsSocket.send("Test.authCallback", "SEND_" + matsSocket.id(6), {});

                chai.assert(authCallbackCalled);
            });

            it('Should invoke authorization callback when room for latency expired', async function () {
                const matsSocket = new MatsSocket("Test", "1.0", ["ws://localhost:8080/"], () => webSocket);

                let authCallbackCalled = false;

                matsSocket.setCurrentAuthorization("Test", Date.now() + 1000, 2000);
                matsSocket.setAuthorizationExpiredCallback(function (event) {
                    authCallbackCalled = true;
                    matsSocket.setCurrentAuthorization("Test", Date.now() + 20000, 0);
                });
                await matsSocket.send("Test.authCallback", "SEND_" + matsSocket.id(6), {});

                chai.assert(authCallbackCalled);
            });
        });
    });

}));

File: mats-websockets/client/dart/test/unit.dart
import 'package:mats_socket/mats_socket.dart';
import 'package:test/test.dart';

import 'logging.dart';
import 'mock_socket_channel.dart';


void main() {
  MockSocketFactory socketFactory;

  configureLogging();

  setUp(() {
    socketFactory = MockSocketFactory.noop();
  });

  group('MatsSocket constructor', () {

    test('Should fail on empty url list', () {
      expect(() => MatsSocket('', '', [], socketFactory), throwsA(TypeMatcher<AssertionError>()));
    });

    test('Should accept a single wsUrl', () {
      MatsSocket('', '', ['ws://test/'], socketFactory);
    });
  });

  group('Authorization', () {
    test('Should invoke authorization callback before making calls', () async {
      var matsSocket = MatsSocket('Test', '1.0', ['ws://localhost:8080/'], socketFactory);
      var authCallbackCalled = false;
      matsSocket.authorizationCallback = (event) {
        authCallbackCalled = true;
        matsSocket.setCurrentAuthorization('Test', DateTime.now().add(Duration(minutes: 1)));
      };

      await matsSocket.send('Test.authCallback', 'SEND_${randomId(6)}', '');

      expect(authCallbackCalled, true);
    });

    test('Should not invoke authorization callback if authorization present', () async {
      var matsSocket = MatsSocket('Test', '1.0', ['ws://localhost:8080/'], socketFactory);
      var authCallbackCalled = false;
      matsSocket.setCurrentAuthorization('Test', DateTime.now().add(Duration(minutes: 1)));
      matsSocket.authorizationCallback = (event) {
        authCallbackCalled = true;  
      };

      await matsSocket.send('Test.authCallback', 'SEND_${randomId(6)}', '');

      expect(authCallbackCalled, false);
    });

    test('Should invoke authorization callback when expired', () async {
      var matsSocket = MatsSocket('Test', '1.0', ['ws://localhost:8080/'], socketFactory);

      var authCallbackCalled = false;

      matsSocket.setCurrentAuthorization('Test', DateTime.now().subtract(Duration(minutes: 1)));
      matsSocket.authorizationCallback = (event) {
        authCallbackCalled = true;
        matsSocket.setCurrentAuthorization('Test', DateTime.now().add(Duration(minutes: 1)));
      };

      await matsSocket.send('Test.authCallback', 'SEND_${randomId(6)}', '');

      expect(authCallbackCalled, true);
    });

    test('Should invoke authorization callback when room for latency expired', () async {
      var matsSocket = MatsSocket('Test', '1.0', ['ws://localhost:8080/'], socketFactory);

      var authCallbackCalled = false;

      matsSocket.setCurrentAuthorization('Test', DateTime.now().subtract(Duration(minutes: 1)), roomForLatency: Duration(minutes: 10));
      matsSocket.authorizationCallback = (event) {
        authCallbackCalled = true;
        matsSocket.setCurrentAuthorization('Test', DateTime.now().add(Duration(minutes: 1)));
      };

      await matsSocket.send('Test.authCallback', 'SEND_' + randomId(6), '');

      expect(authCallbackCalled, true);
    });
  });

  group('Send/Receive', () {
    test('Fail reply on receive failed', () async {
      var socketFactory = MockSocketFactory.withHello((envelope, sink) {
        sink([
          Envelope(
              type: EnvelopeType.RECEIVED,
              subType: EnvelopeSubType.AUTH_FAIL,
              messageSequenceId: envelope.messageSequenceId)
        ]);
      });
      var matsSocket = MatsSocket('Test', '1.0', ['ws://localhost:8080/'], socketFactory);
      matsSocket.setCurrentAuthorization('', DateTime.now().add(Duration(hours: 1)));

      expect(() async => matsSocket.request('Test.failed', 'send', ''),
          throwsException);
    });

    test('Fail reply on reply failed failed', () async {
      var socketFactory = MockSocketFactory.withHello((envelope, sink) {
        sink([
          Envelope(
              type: EnvelopeType.REPLY,
              subType: EnvelopeSubType.REJECT,
              messageSequenceId: envelope.messageSequenceId)
        ]);
      });
      var matsSocket = MatsSocket('Test', '1.0', ['ws://localhost:8080/'], socketFactory);
      matsSocket.setCurrentAuthorization('', DateTime.now().add(Duration(hours: 1)));

      expect(() async => matsSocket.request('Test.failed', 'send', ''),
          throwsException);
    });

    test('Trigger received on Reply', () async {
      var socketFactory = MockSocketFactory.withHello((envelope, sink) {
        sink([
          Envelope(
              type: EnvelopeType.REPLY,
              subType: EnvelopeSubType.RESOLVE,
              messageSequenceId: envelope.messageSequenceId)
        ]);
      });
      var matsSocket = MatsSocket('Test', '1.0', ['ws://localhost:8080/'], socketFactory);
      matsSocket.setCurrentAuthorization('', DateTime.now().add(Duration(hours: 1)));

      var receiveCalled = false;
      await matsSocket.request('Test.failed', 'send', '', (r) { receiveCalled = true; });

      expect(receiveCalled, equals(true));
    });

    test('Handle reply before received', () async {
      var socketFactory = MockSocketFactory.withHello((envelope, sink) {
        sink([
          Envelope(
              type: EnvelopeType.REPLY,
              subType: EnvelopeSubType.RESOLVE,
              messageSequenceId: envelope.messageSequenceId),
          Envelope(
              type: EnvelopeType.RECEIVED,
              subType: EnvelopeSubType.ACK,
              messageSequenceId: envelope.messageSequenceId)
        ]);
      });
      var matsSocket = MatsSocket('Test', '1.0', ['ws://localhost:8080/'], socketFactory);
      matsSocket.setCurrentAuthorization('', DateTime.now().add(Duration(hours: 1)));

      var receiveCalled = false;
      await matsSocket.request('Test.failed', 'send', '', (r) { receiveCalled = true; });

      expect(receiveCalled, equals(true));
    });
  });
}
